lib/ui/channel_buffers.dart

```diff
@@ -6,184 +6,334 @@
+  /// a plugin that notifies the framework of the current value of a
+  /// pressure sensor might leave its size at one (the default), while
+  /// sending messages continually; once the Dart side of the plugin
+  /// registers with the channel, it will immediately receive the most
+  /// up to date value and earlier messages will have been discarded.
+  ///
+  /// Sizes greater than one are appropriate for plugins where every
+  /// message is important. For example, a plugin that itself
+  /// registers with another system that has been buffering events,
+  /// and immediately forwards all the previously-buffered events,
+  /// would likely wish to avoid having any messages dropped on the
+  /// floor. In such situations, it is important to select a size that
+  /// will avoid overflows. It is also important to consider the
+  /// potential for the Dart side to never fully initialize (e.g. if
+  /// the user starts the application, but terminates it soon
+  /// afterwards, leaving time for the platform side of a plugin to
+  /// run but not the Dart side).
   static const int kDefaultBufferSize = 1;
 
+  /// The name of the channel that plugins can use to communicate with the
+  /// channel buffers system.
+  ///
+  /// These messages are handled by [handleMessage].
   static const String kControlChannelName = 'dev.flutter/channel-buffers';
 
-  /// A mapping between a channel name and its associated [_RingBuffer].
-  final Map<String, _RingBuffer<_StoredMessage>?> _messages =
-    <String, _RingBuffer<_StoredMessage>?>{};
-
-  _RingBuffer<_StoredMessage> _makeRingBuffer(int size) {
-    final _RingBuffer<_StoredMessage> result = _RingBuffer<_StoredMessage>(size);
-    result.dropItemCallback = _onDropItem;
-    return result;
-  }
-
-  void _onDropItem(_StoredMessage message) {
-    message.callback(null);
-  }
+  /// A mapping between a channel name and its associated [_Channel].
+  final Map<String, _Channel> _channels = <String, _Channel>{};
 
+  /// Adds a message (`data`) to the named channel buffer (`channel`).
+  ///
+  /// The `callback` argument is a closure that, when called, will send messages
+  /// back to the plugin.
+  ///
   /// Returns true on overflow.
-  bool push(String channel, ByteData? data, PlatformMessageResponseCallback callback) {
-    _RingBuffer<_StoredMessage>? queue = _messages[channel];
-    if (queue == null) {
-      queue = _makeRingBuffer(kDefaultBufferSize);
-      _messages[channel] = queue;
-    }
-    final bool didOverflow = queue.push(_StoredMessage(data, callback));
-    if (didOverflow) {
-      // TODO(gaaclarke): Update this message to include instructions on how to resize
-      // the buffer once that is available to users and print in all engine builds
-      // after we verify that dropping messages isn't part of normal execution.
-      _printDebug('Overflow on channel: $channel. '
-                  'Messages on this channel are being discarded in FIFO fashion. '
```

Comment:
You can't get rid of this warning.  We can't just drop messages without warning developers.  It's particularly an issue with MethodChannels that multiplex multiple messages onto one channel.

Reply:
My understanding is that dropping messages here is part of the expected behaviour of the API. Is that not correct? The documentation talks about how you can legitimately want to only be holding on to the newest message (e.g. because you're a level-trigger API).

We have a policy of only logging messages that are actionable (see style guide). If the messages are not actionable, we shouldn't log the message. I don't know what action one should take with this message.

I agree that if you're dropping messages when you don't expect to be (e.g. you're listening to the wrong channel) then this message would make debugging the issue much easier. Is there some way we can determine when the message is actionable and when it's just logspam?

Reply:
Nope, it's not expected.  Running the engine is an asynchronous operation and we don't provide any mechanism for users to determine if the engine is running before we can send messages.

This code will fail to deliver a message randomly:
```
[engine run];
usleep(rand() * 1000000);
[channel sendMessage:@"hello"];
```

One action that could fix the problem is to chose a more appropriate buffer size, which is mentioned.  If you are using method channels you'd probably want to base that calculus off of the number of unique messages you are sending over that channel.

It also mentions that the engine isn't running which admittedly is a hint about the asynchronous nature of running engines.  With that information they could solve the problem a few other ways.

When I first implemented channel buffers we saw there were all kinds of messages routinely being dropped during the initialization of an engine.  They were not expected.

Reply:
Ok, we should fix the documentation then, because right now it clearly states that there are cases where dropping messages is the appropriate behavior.

Maybe we should show the message for _any_ message that is buffered, if the buffer size hasn't been set? I guess in that case we probably would want to default to a buffer size of zero, too, since we'd be saying all plugins need to set their buffer size if they send messages before the framework is ready.

Reply:
Another option would be to leave the default as now, but have a specific control message to silence the message (and have the warning log say to send this control message).

Reply:
Can you point out the documentation that says dropping messages is appropriate?  One other caveat we haven't discussed is that there are 2 types of channels and the semantics aren't explicit, channels where dropping messages is fine (think sending the gyro readings to the engine) and channels where it isn't (think sending entries in a database).  In other words, is the message interpreted as `set` or `add`?  Is the documentation referring to that?

Buffer size zero by default would be a big breaking change.  I wouldn't recommend that.

The worse case scenario is that users are losing data when they don't expect it.  The print statement is meant to warn users of that scenario.  I worry moving the print statement out of that scenario may cause it to be spammy and less likely heeded when it is appropriate.

However we slice things, not printing out the error despite having set the buffer size wouldn't be ideal since you'd have no idea that you may have chosen poorly for your use case.

> since we'd be saying all plugins need to set their buffer size if they send messages before the framework is ready.  

Remember there is no way to know when the framework is ready clients can't meaningfully know that they may be doing something that may not work.

> have a specific control message to silence the message

That's cool with me.  A bigger change would be if we could know the semantics of the channel and only print out the error if it has `add` semantics instead of `set` semantics.  The ship might have sailed on that though.

Reply:
https://master-api.flutter.dev/flutter/dart-ui/ChannelBuffers/kDefaultBufferSize-constant.html is the documentation I was talking about (I expand on it in this PR). It says "Size 1 implies that you only care about the most recent value." and says size 1 is the default, which implies that dropping messages is fine by default.

We could improve the control channel API in general, making it support a way to specify if it's being used for an event sequence (where dropping is bad) or where it's being used as a level-trigger (where dropping is fine). That would also give us the opportunity to make it use the standard binary format instead of its current inefficent bespoke text format with '\r' as the separator. Let me see what I can do.